File: ios/UsageLimitsKit/Sources/UsageLimitsKit/Providers/XaiBillingParser.swift
import Foundation

public enum XaiBillingParser: Sendable {
    private static let creditsWindowID = "xai-credits"
    private static let monthlyWindowID = "xai-monthly"
    private static let onDemandWindowID = "xai-on-demand"

    // A nominal 30-day month keeps calendar months of different lengths categorised alike.
    private static let billingPeriodSeconds: Int64 = 2_592_000

    // `now` decides whether the reported period is the CURRENT one, which is what makes an
    // absent percentage readable as zero — see below.
    public static func parseCredits(_ payload: [String: Any], now: Date) -> [UsageWindow] {
        let body = config(payload)

        // Two spellings of the period: nested `currentPeriod{type,start,end}` in the credit
        // view, and the flat `usagePeriodType/Start/End` trio the unified-billing view carries
        // beside its monthly figures.
        let period = JSONSupport.object(body, "currentPeriod", "current_period")
        let start = date(period, "start") ?? date(body, "usagePeriodStart", "usage_period_start")
        let end = date(period, "end") ?? date(body, "usagePeriodEnd", "usage_period_end")
        let periodType = (JSONSupport.string(period, "type")
            ?? JSONSupport.string(body, "usagePeriodType", "usage_period_type"))?.lowercased()

        // An ABSENT percentage is a zero when the period is live, and unknown otherwise.
        //
        // xAI's billing message is proto3, and `credit_usage_percent` is an implicit-presence
        // float there: a value of exactly zero is not written to the wire at all. So the one
        // week in which the user has spent nothing — the first week, the week after a reset —
        // arrives with the field missing, and reading that as "no credit window" is how the
        // weekly row vanished from the app the moment it read 100 % remaining. The provider's
        // own web client reads the omitted scalar as zero; so does this, but only when the
        // payload proves it describes the period that contains now. A period in the past, or
        // none at all, is a payload this parser does not understand, and that stays a missing
        // row rather than an invented bar.
        let reported = JSONSupport.double(body, "creditUsagePercent", "credit_usage_percent")
            .flatMap { $0.isFinite ? $0 : nil }
        let periodIsLive: Bool = {
            guard let start, let end else { return false }
            return start <= now && now <= end
        }()
        guard let usedPercent = reported ?? (periodIsLive ? 0.0 : nil) else {
            return []
        }

        var periodSeconds: Int64?
        if let start, let end {
            let span = end.timeIntervalSince(start)
            if span.isFinite, span > 0 {
                // Exact conversion avoids trapping on an out-of-range upstream timestamp.
                periodSeconds = Int64(exactly: span.rounded(.towardZero))
            }
        }

        // Measured length wins; the type preserves weekly/monthly classification when
        // timestamps are missing or unusable.
        if periodSeconds == nil {
            if periodType?.contains("week") == true {
                periodSeconds = 604_800
            } else if periodType?.contains("month") == true {
                periodSeconds = billingPeriodSeconds
            }
        }

        let category = WindowCategory.from(periodSeconds: periodSeconds)
        let label: String
        switch category {
        case .weekly:
            label = "Weekly credits"
        case .monthly:
            label = "Monthly credits"
        default:
            label = "Credits"
        }

        return [
            UsageWindow(
                id: creditsWindowID,
                label: label,
                category: category,
                usedPercent: usedPercent,
                periodSeconds: periodSeconds,
                resetAt: end,
                exhausted: isExhausted(usedPercent)
            )
        ]
    }

    public static func parseBilling(_ payload: [String: Any], now: Date) -> [UsageWindow] {
        let body = config(payload)

        // The unified-billing shape of this view carries the weekly figures too —
        // `creditUsagePercent` with the flat `usagePeriod*` trio — so the weekly row is read
        // from here as well. It costs nothing when both views answer (merge keeps the credit
        // view's copy) and keeps the row when the credit view alone stops answering.
        let weekly = parseCredits(payload, now: now)

        let monthlyLimit = cents(body, "monthlyLimit", "monthly_limit")
        // Every name the production shape carries for spend — see the Kotlin twin for why
        // reading `used` alone rendered a 90 %-spent account as untouched.
        let used = cents(body, "used")
            ?? cents(body, "includedUsed", "included_used")
            ?? cents(body, "totalUsed", "total_used")

        guard monthlyLimit != nil || used != nil else {
            return weekly
        }

        let resetAt = date(body, "billingPeriodEnd", "billing_period_end")

        // Excess spend belongs to the on-demand row. Absent spend stays UNKNOWN: a limit with
        // no spend figure is a window whose percentage the app does not know, not one it has
        // decided is untouched.
        let includedPercent = monthlyLimit.flatMap { limit in
            used.flatMap { percentOf(amount: min($0, limit), total: limit) }
        }

        var windows = weekly + [
            UsageWindow(
                id: monthlyWindowID,
                label: "Monthly included",
                category: .monthly,
                usedPercent: includedPercent,
                periodSeconds: billingPeriodSeconds,
                resetAt: resetAt,
                exhausted: isExhausted(includedPercent)
            )
        ]

        // Without a positive cap, there is no on-demand facility worth showing.
        let onDemandCap = cents(body, "onDemandCap", "on_demand_cap") ?? 0
        if onDemandCap > 0 {
            // Deriving overage needs BOTH a known allowance and known spend; with either
            // missing the row is shown with an unknown percentage rather than omitted.
            let onDemandUsed = cents(body, "onDemandUsed", "on_demand_used")
                ?? monthlyLimit.flatMap { limit in used.map { max(0, $0 - limit) } }
            let onDemandPercent = onDemandUsed.flatMap { percentOf(amount: $0, total: onDemandCap) }

            windows.append(
                UsageWindow(
                    id: onDemandWindowID,
                    label: "On-demand",
                    category: .monthly,
                    usedPercent: onDemandPercent,
                    periodSeconds: billingPeriodSeconds,
                    resetAt: resetAt,
                    exhausted: isExhausted(onDemandPercent)
                )
            )
        }

        return windows
    }

    public static func merge(
        _ credits: [UsageWindow],
        _ billing: [UsageWindow]
    ) -> [UsageWindow] {
        // First occurrence wins so directly reported credits retain precedence if IDs
        // ever collide. Independent endpoint failures need not discard the other view.
        var seen = Set<String>()
        return (credits + billing).filter { seen.insert($0.id).inserted }
    }

    private static func config(_ payload: [String: Any]) -> [String: Any] {
        // Both endpoints wrap their bodies. Flat legacy responses remain valid, but
        // reading only the root silently loses every row from successful HTTP responses.
        JSONSupport.object(payload, "config") ?? payload
    }

    private static func cents(
        _ source: [String: Any]?,
        _ name: String,
        _ alternate: String? = nil
    ) -> Double? {
        // Money may be bare or wrapped in {"val": ...}. Ratios use cents throughout,
        // so no currency conversion or display-unit assumption is necessary.
        let alternateName = alternate ?? name
        if let value = JSONSupport.double(source, name, alternateName) {
            return value
        }
        let wrapped = JSONSupport.object(source, name, alternateName)
        return JSONSupport.double(wrapped, "val")
    }

    private static func date(
        _ source: [String: Any]?,
        _ name: String,
        _ alternate: String? = nil
    ) -> Date? {
        let alternateName = alternate ?? name
        return JSONSupport.date(JSONSupport.string(source, name, alternateName))
            ?? JSONSupport.date(epoch: JSONSupport.int64(source, name, alternateName))
    }

    private static func percentOf(amount: Double, total: Double) -> Double? {
        // Guard the denominator and the result: even valid finite amounts can overflow
        // during percentage calculation, and non-finite bars have no useful meaning.
        guard total > 0 else {
            return nil
        }
        let percentage = amount / total * 100
        return percentage.isFinite ? percentage : nil
    }

    private static func isExhausted(_ usedPercent: Double?) -> Bool {
        guard let usedPercent else {
            return false
        }
        return usedPercent >= 100
    }
}

File: app/src/main/kotlin/com/usagelimits/providers/xai/XaiBillingParser.kt
package com.usagelimits.providers.xai

import com.usagelimits.core.model.UsageWindow
import com.usagelimits.core.model.WindowCategory
import com.usagelimits.core.network.JsonSupport
import com.usagelimits.core.time.Instants
import kotlinx.serialization.json.JsonObject

/**
 * Turns the two xAI billing payloads into normalised windows.
 *
 * xAI is the odd provider out: it does not report quota as a percentage of a rate limit, it
 * reports spend. Two views exist and they describe different periods, so the app reads both:
 *
 *  - `/v1/billing?format=credits` — the weekly credit view, the one endpoint that does hand
 *    back a ready-made percentage.
 *  - `/v1/billing` — the monthly view, where **every amount is an integer number of cents**.
 *
 * Nothing here converts or displays currency. Percentages are ratios of two cent amounts, so
 * the unit cancels out; the app never shows the user a money value it might get wrong.
 *
 * The two views are parsed separately and joined by [merge] so that losing one endpoint costs
 * its rows rather than the whole account.
 */
object XaiBillingParser {

    const val CREDITS_WINDOW_ID = "xai-credits"
    const val MONTHLY_WINDOW_ID = "xai-monthly"
    const val ON_DEMAND_WINDOW_ID = "xai-on-demand"

    /**
     * The billing endpoint reports period *stamps*, not a period length, and a real calendar
     * month is 28-31 days. 30 days is the nominal billing month and sits inside the monthly
     * band [WindowCategory.fromPeriodSeconds] recognises, so the window is categorised as
     * monthly whatever month it happens to be.
     */
    private const val BILLING_PERIOD_SECONDS = 2_592_000L

    /**
     * Unwraps the `config` envelope both billing endpoints wrap their payload in.
     *
     * Upstream reads `payload.config` (Management Center
     * src/features/quota/providers/xai/data.ts:97). Older or partial responses have been seen
     * flat, so the root is accepted as a fallback rather than requiring the envelope.
     */
    private fun config(payload: JsonObject): JsonObject =
        JsonSupport.obj(payload, "config") ?: payload

    /**
     * Reads a money field, which xAI reports either as a bare number or wrapped as
     * `{"val": 10000}`.
     *
     * Management Center normalises the same two shapes in normalizeXaiCentValue
     * (src/utils/quota/builders.ts:374). Reading only the bare form yields null for a wrapped
     * value, which would silently drop the whole billing view.
     */
    private fun cents(source: JsonObject?, vararg names: String): Double? {
        JsonSupport.double(source, *names)?.let { return it }
        val wrapped = JsonSupport.obj(source, *names) ?: return null
        return JsonSupport.double(wrapped, "val")
    }

    /**
     * Parses the weekly credit view.
     *
     * [nowMs] is part of the parser contract the four providers share; xAI states the period
     * end absolutely, so nothing here currently needs it.
     */
    @Suppress("UNUSED_PARAMETER")
    fun parseCredits(payload: JsonObject, nowMs: Long): List<UsageWindow> {
        // No percentage means no credit window at all — better one missing row than a bar
        // drawn from an assumed zero.
        val config = config(payload)
        val usedPercent = JsonSupport.double(config, "creditUsagePercent", "credit_usage_percent")
            ?.takeIf { it.isFinite() }
            ?: return emptyList()

        val period = JsonSupport.obj(config, "currentPeriod", "current_period")
        val startMs = Instants.parse(JsonSupport.string(period, "start"))
        val endMs = Instants.parse(JsonSupport.string(period, "end"))
        val periodType = JsonSupport.string(period, "type")

        // The measured span takes precedence so a change upstream reclassifies itself instead
        // of mislabelling a fortnight as a week. Without a usable span, the reported type can
        // still identify the period; without either, OTHER avoids claiming a known length.
        val periodSeconds = if (startMs != null && endMs != null && endMs > startMs) {
            (endMs - startMs) / 1000
        } else {
            when {
                periodType?.contains("week", ignoreCase = true) == true -> 604_800L
                periodType?.contains("month", ignoreCase = true) == true -> BILLING_PERIOD_SECONDS
                else -> null
            }
        }
        val category = WindowCategory.fromPeriodSeconds(periodSeconds)

        return listOf(
            UsageWindow(
                id = CREDITS_WINDOW_ID,
                label = when (category) {
                    WindowCategory.WEEKLY -> "Weekly credits"
                    WindowCategory.MONTHLY -> "Monthly credits"
                    else -> "Credits"
                },
                category = category,
                usedPercent = usedPercent,
                periodSeconds = periodSeconds,
                resetAt = endMs,
                exhausted = isExhausted(usedPercent),
            ),
        )
    }

    /**
     * Parses the monthly spend view into an included-allowance window and, where the account
     * has one, a separate on-demand window.
     *
     * They are two bars, not one, because they run out independently: an account can have
     * burned its whole monthly allowance and still be able to spend on demand.
     *
     * [nowMs] is unused for the same reason as in [parseCredits].
     */
    @Suppress("UNUSED_PARAMETER")
    fun parseBilling(payload: JsonObject, nowMs: Long): List<UsageWindow> {
        val config = config(payload)
        val monthlyLimit = cents(config, "monthlyLimit", "monthly_limit")
        // Every name the production shape carries for spend. Reading `used` alone meant a
        // payload whose spend arrived as `includedUsed` — the pinned production shape — read
        // as ZERO spend: "0 % used / 100 % remaining", healthy green, for an account 90 %
        // through its allowance.
        val used = cents(config, "used")
            ?: cents(config, "includedUsed", "included_used")
            ?: cents(config, "totalUsed", "total_used")
        // Neither figure present means this is not a billing payload we understand.
        if (monthlyLimit == null && used == null) return emptyList()

        val resetAt = Instants.parse(
            JsonSupport.string(config, "billingPeriodEnd", "billing_period_end"),
        )

        val windows = mutableListOf<UsageWindow>()

        // Spend past the allowance is on-demand spend, so the included bar stops at 100 %.
        // And spend that is ABSENT stays unknown: a limit with no spend figure is a window
        // whose percentage the app does not know, never one it has decided is untouched.
        val includedPercent = monthlyLimit?.let { limit ->
            used?.let { percentOf(minOf(it, limit), limit) }
        }
        windows += UsageWindow(
            id = MONTHLY_WINDOW_ID,
            label = "Monthly included",
            category = WindowCategory.MONTHLY,
            usedPercent = includedPercent,
            periodSeconds = BILLING_PERIOD_SECONDS,
            resetAt = resetAt,
            exhausted = isExhausted(includedPercent),
        )

        // A zero or absent cap means the account cannot spend on demand at all; showing an
        // empty bar for a facility that does not exist would just be noise.
        val onDemandCap = cents(config, "onDemandCap", "on_demand_cap") ?: 0.0
        if (onDemandCap > 0.0) {
            // Deriving overage needs BOTH a known allowance and known spend; with either
            // missing the row is shown with an unknown percentage rather than omitted or
            // invented as zero.
            val onDemandUsed = cents(config, "onDemandUsed", "on_demand_used")
                ?: monthlyLimit?.let { limit -> used?.let { maxOf(0.0, it - limit) } }
            val onDemandPercent = onDemandUsed?.let { percentOf(it, onDemandCap) }
            windows += UsageWindow(
                id = ON_DEMAND_WINDOW_ID,
                label = "On-demand",
                category = WindowCategory.MONTHLY,
                usedPercent = onDemandPercent,
                periodSeconds = BILLING_PERIOD_SECONDS,
                resetAt = resetAt,
                exhausted = isExhausted(onDemandPercent),
            )
        }

        return windows
    }

    /**
     * Joins the two views, credits first.
     *
     * The parsers' ids are currently disjoint by construction, so de-duplication guards
     * against a future id collision rather than removing any rows today. First occurrence
     * wins, so the credit view — the only one that reports a percentage directly — takes
     * precedence if a collision is introduced.
     */
    fun merge(credits: List<UsageWindow>, billing: List<UsageWindow>): List<UsageWindow> {
        val seen = mutableSetOf<String>()
        return (credits + billing).filter { seen.add(it.id) }
    }

    /**
     * The single division point in this file.
     *
     * Every ratio goes through it so a zero or missing denominator can only ever produce a
     * null percentage — never NaN or Infinity, which would render as a nonsense bar and would
     * compare as "healthy" against the severity thresholds.
     */
    private fun percentOf(amount: Double, total: Double): Double? =
        if (total > 0.0) (amount / total * 100.0).takeIf { it.isFinite() } else null

    private fun isExhausted(usedPercent: Double?): Boolean =
        usedPercent != null && usedPercent >= 100.0
}
